hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -136,67 +180,68 @@ withReliability ::
+  -- | All parties' identifiers.
+  Vector Party ->
   -- | Underlying network component providing consuming and sending channels.
   NetworkComponent m (Authenticated (ReliableMsg (Heartbeat msg))) (ReliableMsg (Heartbeat msg)) a ->
   NetworkComponent m (Authenticated (Heartbeat msg)) (Heartbeat msg) a
-withReliability tracer me otherParties withRawNetwork callback action = do
-  ackCounter <- newTVarIO $ replicate (length allParties) 0
-  sentMessages <- newTVarIO IMap.empty
+withReliability tracer MessagePersistence{saveAcks, loadAcks, appendMessage, loadMessages} me allParties withRawNetwork callback action = do
+  acksCache <- loadAcks >>= newTVarIO
   resendQ <- newTQueueIO
   let ourIndex = fromMaybe (error "This cannot happen because we constructed the list with our party inside.") (findPartyIndex me)
   let resend = writeTQueue resendQ
-  withRawNetwork (reliableCallback ackCounter sentMessages resend ourIndex) $ \network@Network{broadcast} -> do
+  withRawNetwork (reliableCallback acksCache resend ourIndex) $ \network@Network{broadcast} -> do
     withAsync (forever $ atomically (readTQueue resendQ) >>= broadcast) $ \_ ->
-      reliableBroadcast ourIndex ackCounter sentMessages network
+      reliableBroadcast ourIndex acksCache network
  where
-  allParties = fromList $ sort $ me : otherParties
-
-  reliableBroadcast ourIndex ackCounter sentMessages Network{broadcast} =
+  reliableBroadcast ourIndex acksCache Network{broadcast} =
     action $
       Network
         { broadcast = \msg ->
             case msg of
               Data{} -> do
-                ackCounter' <- atomically $ incrementCountersFor msg
-                traceWith tracer (BroadcastCounter ourIndex ackCounter')
-                broadcast $ ReliableMsg ackCounter' msg
+                newAckCounter <- incrementAckCounter
+                appendMessage msg
+                saveAcks newAckCounter
+                traceWith tracer (BroadcastCounter ourIndex newAckCounter)
+                broadcast $ ReliableMsg newAckCounter msg
               Ping{} -> do
-                acks <- readTVarIO ackCounter
+                acks <- readTVarIO acksCache
                 traceWith tracer (BroadcastPing ourIndex acks)
                 broadcast $ ReliableMsg acks msg
         }
    where
-    incrementCountersFor msg = do
-      acks <- readTVar ackCounter
-      writeTVar ackCounter $ constructAcks acks ourIndex
-      modifyTVar' sentMessages (insertNewMsg msg)
-      readTVar ackCounter
+    incrementAckCounter = atomically $ do
+      acks <- readTVar acksCache
+      let newAcks = constructAcks acks ourIndex
+      writeTVar acksCache newAcks
+      pure newAcks
 
-  reliableCallback ackCounter sentMessages resend ourIndex (Authenticated (ReliableMsg acks msg) party) = do
+  reliableCallback acksCache resend ourIndex (Authenticated (ReliableMsg acks msg) party) = do
     if length acks /= length allParties
-      then ignoreMalformedMessages
+      then pure ()
```

Comment:
We should log the fact we received a f**ed up message from a peer because this means our Head cannot work or someone is trying to screw us.